src/libopensc/card-itacns.c

```diff
@@ -309,6 +356,17 @@ static int itacns_set_security_env(sc_card_t *card,
 		(card->type == SC_CARD_TYPE_ITACNS_CIE_V1 ? 0x30 : 0x03));
 	LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
 
+	if (card->type == SC_CARD_TYPE_ITACNS_CNS_IDEMIA_2021) {
+		/* We need to select this path in order to be able to perform security
+		 * operations, without it decipher operation would fail after sending
+		 * the data with a 'Reference data not usable' error */
```

Comment:
According to https://www.agid.gov.it/sites/default/files/repository_files/documentazione/filesystemcns_20160610.pdf
 you are selecting some file in DF_DS, which is presumably your signature key. I think what you actually want, is to add this key/path to the card's profile, see:
https://github.com/OpenSC/OpenSC/blob/05ec8c7fe785a2b9aeaac1164adb349df42b7f80/src/libopensc/pkcs15-itacns.c#L848-L854

Reply:
Mh, could be but I'm not sure it's that, as that path isn't really a file:

```
P:2964129; T:0x140541533505536 01:41:44.822 [opensc-pkcs11] pkcs15-itacns.c:206:itacns_add_cert: called
P:2964129; T:0x140541533505536 01:41:44.823 [opensc-pkcs11] pkcs15-cert.c:381:sc_pkcs15_read_certificate: called
P:2964129; T:0x140541533505536 01:41:44.823 [opensc-pkcs11] pkcs15.c:2355:sc_pkcs15_read_file: called
P:2964129; T:0x140541533505536 01:41:44.823 [opensc-pkcs11] pkcs15.c:2356:sc_pkcs15_read_file: path=3f0014009002, index=0, count=-1
P:2964129; T:0x140541533505536 01:41:44.823 [opensc-pkcs11] card.c:473:sc_lock: called
P:2964129; T:0x140541533505536 01:41:44.823 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.823 [opensc-pkcs11] card.c:844:sc_select_file: called; type=2, path=3f0014009002
P:2964129; T:0x140541533505536 01:41:44.824 [opensc-pkcs11] card-itacns.c:589:itacns_select_file: called
P:2964129; T:0x140541533505536 01:41:44.824 [opensc-pkcs11] apdu.c:550:sc_transmit_apdu: called
P:2964129; T:0x140541533505536 01:41:44.824 [opensc-pkcs11] card.c:473:sc_lock: called
P:2964129; T:0x140541533505536 01:41:44.824 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.824 [opensc-pkcs11] apdu.c:515:sc_transmit: called
P:2964129; T:0x140541533505536 01:41:44.825 [opensc-pkcs11] apdu.c:363:sc_single_transmit: called
P:2964129; T:0x140541533505536 01:41:44.825 [opensc-pkcs11] apdu.c:367:sc_single_transmit: CLA:0, INS:A4, P1:8, P2:0, data(4) 0x7ffed1b84062
P:2964129; T:0x140541533505536 01:41:44.825 [opensc-pkcs11] reader-pcsc.c:325:pcsc_transmit: reader 'Alcor Micro AU9560 00 00'
P:2964129; T:0x140541533505536 01:41:44.826 [opensc-pkcs11] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (10 bytes):
00 A4 08 00 04 14 00 90 02 00 ..........

P:2964129; T:0x140541533505536 01:41:44.826 [opensc-pkcs11] reader-pcsc.c:244:pcsc_internal_transmit: called
P:2964129; T:0x140541533505536 01:41:44.845 [opensc-pkcs11] reader-pcsc.c:335:pcsc_transmit: 
Incoming APDU (57 bytes):
6F 35 81 02 00 00 82 03 38 00 00 83 02 90 02 85 o5......8.......
01 01 86 09 FF 10 10 FF FF FF FF 10 FF CB 18 FF ................
FF 03 04 03 04 FF FF FF FF FF FF FF FF FF FF FF ................
FF FF FF FF FF FF FF 90 00                      .........

P:2964129; T:0x140541533505536 01:41:44.845 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.846 [opensc-pkcs11] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.846 [opensc-pkcs11] card.c:523:sc_unlock: called
P:2964129; T:0x140541533505536 01:41:44.846 [opensc-pkcs11] iso7816.c:355:iso7816_process_fci:   bytes in file: 0
P:2964129; T:0x140541533505536 01:41:44.846 [opensc-pkcs11] iso7816.c:366:iso7816_process_fci:   shareable: no
P:2964129; T:0x140541533505536 01:41:44.846 [opensc-pkcs11] iso7816.c:386:iso7816_process_fci:   type: DF
P:2964129; T:0x140541533505536 01:41:44.846 [opensc-pkcs11] iso7816.c:387:iso7816_process_fci:   EF structure: 0
P:2964129; T:0x140541533505536 01:41:44.847 [opensc-pkcs11] iso7816.c:388:iso7816_process_fci:   tag 0x82: 0x38
P:2964129; T:0x140541533505536 01:41:44.847 [opensc-pkcs11] iso7816.c:419:iso7816_process_fci:   file identifier: 0x9002
P:2964129; T:0x140541533505536 01:41:44.847 [opensc-pkcs11] card-itacns.c:595:itacns_select_file: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.847 [opensc-pkcs11] card.c:879:sc_select_file: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.847 [opensc-pkcs11] card.c:629:sc_read_binary: called; 1024 bytes at index 0
P:2964129; T:0x140541533505536 01:41:44.847 [opensc-pkcs11] card.c:473:sc_lock: called
P:2964129; T:0x140541533505536 01:41:44.848 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.848 [opensc-pkcs11] apdu.c:550:sc_transmit_apdu: called
P:2964129; T:0x140541533505536 01:41:44.848 [opensc-pkcs11] card.c:473:sc_lock: called
P:2964129; T:0x140541533505536 01:41:44.848 [opensc-pkcs11] card.c:513:sc_lock: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.848 [opensc-pkcs11] apdu.c:515:sc_transmit: called
P:2964129; T:0x140541533505536 01:41:44.849 [opensc-pkcs11] apdu.c:363:sc_single_transmit: called
P:2964129; T:0x140541533505536 01:41:44.849 [opensc-pkcs11] apdu.c:367:sc_single_transmit: CLA:0, INS:B0, P1:0, P2:0, data(0) (nil)
P:2964129; T:0x140541533505536 01:41:44.849 [opensc-pkcs11] reader-pcsc.c:325:pcsc_transmit: reader 'Alcor Micro AU9560 00 00'
P:2964129; T:0x140541533505536 01:41:44.849 [opensc-pkcs11] reader-pcsc.c:326:pcsc_transmit: 
Outgoing APDU (5 bytes):
00 B0 00 00 FF .....

P:2964129; T:0x140541533505536 01:41:44.849 [opensc-pkcs11] reader-pcsc.c:244:pcsc_internal_transmit: called
P:2964129; T:0x140541533505536 01:41:44.856 [opensc-pkcs11] reader-pcsc.c:335:pcsc_transmit: 
Incoming APDU (2 bytes):
69 86 i.

P:2964129; T:0x140541533505536 01:41:44.857 [opensc-pkcs11] apdu.c:382:sc_single_transmit: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.857 [opensc-pkcs11] apdu.c:539:sc_transmit: returning with: 0 (Success)
P:2964129; T:0x140541533505536 01:41:44.857 [opensc-pkcs11] card.c:523:sc_unlock: called
P:2964129; T:0x140541533505536 01:41:44.857 [opensc-pkcs11] iso7816.c:128:iso7816_check_sw: Command not allowed (no current EF)
P:2964129; T:0x140541533505536 01:41:44.857 [opensc-pkcs11] iso7816.c:162:iso7816_read_binary: Check SW error: -1209 (Not allowed)
P:2964129; T:0x140541533505536 01:41:44.858 [opensc-pkcs11] card.c:523:sc_unlock: called
P:2964129; T:0x140541533505536 01:41:44.858 [opensc-pkcs11] card.c:662:sc_read_binary: returning with: -1415 (EF offset too large)
P:2964129; T:0x140541533505536 01:41:44.858 [opensc-pkcs11] card.c:523:sc_unlock: called
P:2964129; T:0x140541533505536 01:41:44.859 [opensc-pkcs11] pkcs15.c:2460:sc_pkcs15_read_file: returning with: -1415 (EF offset too large)
P:2964129; T:0x140541533505536 01:41:44.859 [opensc-pkcs11] pkcs15-cert.c:388:sc_pkcs15_read_certificate: Unable to read certificate file.: -1415 (EF offset too large)
P:2964129; T:0x140541533505536 01:41:44.859 [opensc-pkcs11] pkcs15-itacns.c:231:itacns_add_cert: Could not read X.509 certificate: -1415 (EF offset too large)
P:2964129; T:0x140541533505536 01:41:44.859 [opensc-pkcs11] pkcs15-itacns.c:719:itacns_check_and_add_keyset: Could not add certificate: -1415 (EF offset too large)
P:2964129; T:0x140541533505536 01:41:44.859 [opensc-pkcs11] pkcs15-itacns.c:844:itacns_init: Could not add CNS0: -1415 (EF offset too large)
P:2964129; T:0x140541533505536 01:41:44.860 [opensc-pkcs11] pkcs15-syn.c:174:sc_pkcs15_bind_synthetic: trying PIV-II
```

Reply:
```
OpenSC [3F00/1400]> info 9002

Dedicated File  ID 9002

File path:               3F00/1400/9002
File size:               0 bytes
ACL for SELECT:          N/A
ACL for LOCK:            N/A
ACL for DELETE:          N/A
ACL for CREATE:          N/A
ACL for REHABILITATE:    N/A
ACL for INVALIDATE:      N/A
ACL for LIST FILES:      N/A
ACL for CRYPTO:          N/A
ACL for DELETE SELF:     N/A
Type attributes:         38 00 00
Proprietary attributes:  01
Security attributes:     FF 10 10 FF FF FF FF 10 FF
Life cycle:              LCSB is not present

OpenSC [3F00/1400]> info 9001

Dedicated File  ID 9001

File path:               3F00/1400/9001
File size:               0 bytes
ACL for SELECT:          N/A
ACL for LOCK:            N/A
ACL for DELETE:          N/A
ACL for CREATE:          N/A
ACL for REHABILITATE:    N/A
ACL for INVALIDATE:      N/A
ACL for LIST FILES:      N/A
ACL for CRYPTO:          N/A
ACL for DELETE SELF:     N/A
Type attributes:         38 00 00
Proprietary attributes:  01
Security attributes:     FF 00 00 FF FF 00 FF 00 FF
Life cycle:              LCSB is not present
```

Reply:
read `itacns_check_and_add_keyset()` and try to understand what it does... with your additional select, you're selecting the private key, which, I assume, is also done when you use `itacns_check_and_add_keyset()` properly.

Reply:
Bah, I was sure to have tried that without success, but apparently I forgot something and you're right.